### ISSUE TITLE
Conditional requirements

### ISSUE THREAD

**Issue by sminnee** [composer/composer#4168]
Sometimes you have modules that are designed to connect two other modules. An example that we've come across is in the interaction between themes and modules:

 * A theme for a CMS, e.g. "silverstripe-themes/simple"
 * An add-on module, e.g. "silverstripe/forum"
 * An additional theme that ensures that module is styled consistently for that theme, "silverstripe-themes/simple_forum"

What I would like to do is say if "silverstripe-themes/simple" *and* "silverstripe/forum" are installed, then "silverstripe-themes/simple_forum" should also be a requirement.

One way of describing this would be conditional requirements, maybe something like this:

    {
      "name": "silverstripe-themes/simple",
      ...
      "conditional": [
        {
          "if-installed": {
            "silverstripe/forum": "*",
           },
           "require": {
             "silvestripe-themes/simple_forum": "*"
           }
        }
      ]
    }

I'm not familiar with how the dependency solver works. One way to do it would be to iterate multiple passes over the solver: if the if-installed conditions were met, then add those requirements to the definition of the relevant package and re-execute the solver. I don't know enough about its internals to determine the feasibility of that approach.

Is this something that other people would find of value? Is this something that would be feasible to implement?

A few external refs:

 * It seem that Solaris' package manager supports this: http://docs.oracle.com/cd/E36784_01/html/E36856/gluld.html
 * It came up in the NPM issue tracker and they suggested using a post-install script (although you would to support post-install scripts for included packages for that to work): https://github.com/npm/npm/issues/7340

**Comment by sminnee** [comment]
OK, so it looks like #1193 has discussed package updates. Perhaps if the feature discussed in that issue as implemented, this could be achieved with a script like this?

    "scripts": {
        "post-package-update": "composer info --installed silverstripe/forum &> /dev/null && composer require silvestripe-themes/simple_forum"
     }

One thing to note there is that it would be nice to be able to say "execute composer" without knowing where composer was installed. Maybe $COMPOSER_SELF could point to the command that you can use to execute composer, or there could be a way of executing composer command objects, similar to how you can call static methods on objects.

**Comment by slbmeh** [comment]
Duplicate of #751?

**Comment by mcd-php** [comment]
#### If you need it, your design is bad.

Do you really want combinatorial explosion between modules like forum, shop, blog, gallery etc, and themes like simple, complex, yellow, green, red etc. ? If I want to contribute theme for your project, do I need to enumerate all modules and create sub-themes for them ? If I want to contribute module, do I need to enumerate all themes and create sub-themes for all of them ? On conditions like this, I will not contribute at all, unless absolutely rat-cornered, and your project is doomed.

Please define finite set of visual features ( header, breadcrumbs, tag cloud, sub-title, picture with text etc. ) and make sure you only use them, and not a wild HTML in views. Then themes should only restyle defined blocks.

#### Then you really need a bridge package between the two, 
... you make it depend on both bridged ones and list it as main dependency, and others are auto-installed. Or otherwise, you have some of dependencies omitted from description - this may be a subtle bug, but just document it and go on.

For example, I code in Yii2 and come across some nice validators library written for Zend2, say "foomatic\biz-valid", but not at all need any validators from base ZF2. Then I make separate component "acme\yii2-zf2-validators" and make it depend on Yii2 and `zendframework/zend-validator`, since the latter has `Zend\Validator\ValidatorInterface` I must typehint for.

But my custom package is not dependent on "foomatic\biz-valid", since it can perfectly go without it specifically, but with any other ZF2 validator. They all three are only depended by my main application.

```gv
digraph "acme/main-app" {
  "acme/main-app" -> "acme/yii2-zf2-validators"
  "acme/main-app" -> "yiisoft/yii2"
  "acme/main-app" -> "foomatic/biz-valid"
  "acme/yii2-zf2-validators" -> "zendframework/zend-validator"
  "acme/yii2-zf2-validators" -> "yiisoft/yii2"
  "foomatic/biz-valid" -> "zendframework/zend-validator"
  "acme/yii2-zf2-validators" -> "foomatic/biz-valid" [ color="red" style="dashed" label="NO" ]
}
```
![example dependency graph](http://i.imgur.com/ATiSfZv.png)

**Comment by assertchris** [comment]
So if I want to make an app that uses a combination of Yii + Zend Validators and a combination of Yii + Monolog I would need to name each different library in the name of the composite package (yii2-zf2-validators-monolog), or have different packages requiring the same sub-dependencies (with potential version conflicts)..?

**Comment by mcd-php** [comment]
Hm, we strayed from main point: conditional package install is misfeature, leading to misdesign, grief and torn hair.

My point was: the package candidate to being conditional (the "bridge" or "adapter" or "suggested" one) should become main dependency, and the "ends" it stands on should become secondary dependencies.

No you should not name package after all it's dependencies. Instead, write individual "adapter" or "wrapper" packages just like you want to open-source them, even if you don't. Then name them after your team and both bridged packages, like "acme/yii2-log-monolog". Then use them from your main app(s).

About version clash: you control all yours "acme" packages and can boost external dependencies versions in orchestrated manner. The only real clash is true transient dependencies: then two external packages depend on incompatible versions of same library.

About combinatorial explosion: just don't design this way, make all dimensions of cube talk to some solid interface. For website extensions and themes - define finite set of widgets. For accounting system, having documents and reports, avoid co-dependence by introducing registries, write to them on document posting and read on reporting, so you can add new documents and reports without touching counterparts, i.e. making all reports read documents of new kind.

P.S. I found out how to avoid red arrows (non-specified dependencies) - just wrap the offending package in separate adapter package, not in application.

```gv
digraph "acme/main-app" {

  node [ shape=box ]

  "acme/main-app" -> "yiisoft/yii2"
  "acme/main-app" -> "acme/yii2-log-monolog"
  "acme/main-app" -> "acme/yii2-foomatic-biz-valid"
  "acme/yii2-foomatic-biz-valid" -> "acme/yii2-zf2-validators"
  "acme/yii2-foomatic-biz-valid" -> "foomatic/biz-valid"
  "acme/yii2-zf2-validators" -> "zendframework/zend-validator"
  "acme/yii2-zf2-validators" -> "yiisoft/yii2"
  "acme/yii2-log-monolog" -> "monolog/monolog"
  "acme/yii2-log-monolog" -> "yiisoft/yii2"
  "foomatic/biz-valid" -> "zendframework/zend-validator"
}
```
![bridge packages](http://i.imgur.com/61bQrS7.png)

In this scheme, you can easily open-source all your "acme" packages ex for "main", since they have no business logic and will be benefitable for everyone. Or at least you can reuse them in all your projects with same base framework, yii2, and not repeat yourself in all applications wrapping foreign packages again and again..

**Comment by assertchris** [comment]
Your diagram looks like a nightmare. Ok, I get that you have thought about this for ages but consider how complex this design is to explain and keep in mind. And bumping the dependency versions, across the board, is a pain for any SemVer-compliant, popular project. Is it possible that this could be over-engineering?

**Comment by sminnee** [comment]
@mcd-php "if you need it, your design is bad" feels a little presumptuous. Defining a finite set of visual features and using only those isn't really practical for our project, and although we could get into a big discussion about it, that's beside the point—we're here to talk about Composer, not about my project.

I acknowledge the combinatorial issues you've pointed out here, but I don't think the design here is so bad that it would render conditional dependencies an unnecessary feature, especially since they have been discussed in many other package management systems, including Composer.

----

@slbmeh conditional dependencies and alternative dependencies are related, but separate features. It would probably be worth considering their implementation & configuration together.

**Comment by mcd-php** [comment]
@sminnee No one can discuss Composer without discussing example dependency trees, and so the real projects.

Short: then you have `modules that are designed to connect two other modules`, either make them primary dependencies (`my-yii2-project` -> `yii2-monolog`->`yiisoft/yii2`,`monolog/monolog`), or rethink your tree, or code to interfaces and not to implementations - if `monolog` implements PSR-3 logging interface both in and out, it opens the whole range of opportunities.

Long: 

The one and only valid reason to depend on a package is using the elements of it's code in your package's code: creating new instances, subclassing, typehinting method arguments and returns, calling statically.

In writing a bridge module, you need everything from modules being bridged: interfaces to implement, input classes to typehint arguments, output classes to typehint returns, etc.

And the most important: this facts of usage do not blink because of anything in external packages, they are final and persistent. If you use class from forum and main theme, you depend on them.

Since bridge modules depend on main module, you may list only the bridges in the application, and base modules will be pulled in by "bridges".

![Imgur](http://i.imgur.com/XSBxdet.png?1)

**Comment by sminnee** [comment]
I spoke with @Seldaek and because this is an edge-case requirement that would significantly increase composer's complexity and maintenance-cost, I think that this would be best attempted as an add-on module. I'll let you know if we get anywhere. Until then, ticket closed!
